File: Problem25.java
//Time Complexity: O(n)
//Space Complexity: O(1)
//Did it run on leetcode: Yes
//Problems faced any: No


public class Problem25 {
    public int[] productExceptSelf(int[] nums) {
        if(nums==null || nums.length==0)
            return new int[0];

        int[] output = new int[nums.length];

        //compute left array
        int product = 1;
        for(int i=0; i<nums.length; i++){
            output[i] = product;
            product *= nums[i];
        }

        //compute right array
        //here we are actually merging both of them(left and right) so using only one array as output
        product = 1;
        for(int i=nums.length-1; i>=0; i--){
            output[i] *= product;
            product *= nums[i];
        }

        return output;

    }
}

File: Problem25.py
def productExceptSelf(array):
    left= [1] * len(array)
    right= [1]* len(array)

# product of left side of the array
    for i in range(1, len(nums)):
        left[i] = left[i-1]*nums[i-1]
#product of right side of the array
    for j in range(len(nums)-2, -1,-1):
        right[j] = right[j+1]* nums[i+1]
#combining the prodoct of left and right
    res = [1] * len(nums)
    for i in range(len(nums)):
        res[i] = left[i] * right[i]
    return res #contains the product of except self results.

"""
Time Complexity: O(n)
Space Complexity: O(n)

"""
 
